File: backend/app/schemas/translation.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, field_validator


class OllamaRequestOptions(BaseModel):
    think: str | bool = False
    options: dict[str, Any] | None = None

    @field_validator("think", mode="before")
    @classmethod
    def validate_think(cls, value: Any) -> Any:
        if isinstance(value, bool) or isinstance(value, str):
            return value
        raise ValueError("think must be a string or boolean")


class TranslationRequest(OllamaRequestOptions):
    text: str
    source_lang: str = "ja"
    target_lang: str = "ko"
    translate_scope: Literal["first_page", "current_page", "all_pages"] = "first_page"
    page_index: int = 0
    style: str = "webnovel"
    honorific_policy: str = "preserve"
    preserve_names: bool = True
    use_glossary: bool = True
    use_cache: bool = True
    stream: bool = False


class PageTranslateRequest(OllamaRequestOptions):
    source_lang: str = "ja"
    target_lang: str = "ko"
    translate_scope: Literal["current_page"] = "current_page"
    style: str = "webnovel"
    honorific_policy: str = "preserve"
    preserve_names: bool = True
    use_glossary: bool = True
    use_cache: bool = True
    stream: bool = False
    force: bool = False


class TranslationChunkResponse(BaseModel):
    index: int
    source_lang: str
    target_lang: str
    status: str


class TranslationResponse(BaseModel):
    job_id: int
    source_type: str = "pasted_text"
    source_lang: str
    target_lang: str
    current_page_index: int
    total_pages: int
    has_next_page: bool
    translated_text: str
    model: str
    prompt_version: str
    style: str
    elapsed_ms: int
    cache_hit: bool
    chunks: list[TranslationChunkResponse]


class TranslationHistoryItem(BaseModel):
    job_id: int
    source_site: str
    source_url: str | None
    source_title: str | None
    source_author: str | None
    source_work_id: str | None
    source_fetched_at: datetime | None
    source_preview: str
    translated_preview: str | None
    source_lang: str
    target_lang: str
    model_name: str
    prompt_version: str
    ollama_think: str | None
    ollama_options_json: str | None
    style: str
    honorific_policy: str
    preserve_names: bool
    status: str
    total_pages: int
    total_chunks: int
    completed_chunks: int
    failed_chunks: int
    elapsed_ms: int | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime


class TranslationPageHistory(BaseModel):
    id: int
    page_index: int
    page_title: str | None
    source_text: str
    translated_text: str | None
    status: str
    total_chunks: int
    completed_chunks: int
    failed_chunks: int
    elapsed_ms: int | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime


class TranslationChunkHistory(BaseModel):
    id: int
    page_id: int
    page_index: int | None
    chunk_index: int
    source_lang: str
    target_lang: str
    source_text: str
    translated_text: str | None
    context_before: str | None
    context_after: str | None
    status: str
    retry_count: int
    prompt_used: str | None
    raw_model_response: str | None
    elapsed_ms: int | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime


class TranslationDetailResponse(TranslationHistoryItem):
    original_text: str
    translated_text: str | None
    pages: list[TranslationPageHistory]
    chunks: list[TranslationChunkHistory]

File: frontend/types/translation.ts
export type SourceLanguage = "auto" | "ja" | "zh-TW" | "zh-CN" | "en";
export type TargetLanguage = "ko";
export type TranslateScope = "first_page" | "current_page" | "all_pages";
export type InputMode = "url" | "text";
export type ViewerMode = "both" | "translation";
export type TranslationStyle = "webnovel" | "lightnovel" | "literal" | "natural";
export type HonorificPolicy = "preserve" | "naturalize" | "omit";
export type PageStatus = "pending" | "pending_translation" | "in_progress" | "fetched" | "completed" | "failed";
export type TranslationStatus =
  | "idle"
  | "ready"
  | "fetched"
  | "pending_translation"
  | "translating"
  | "completed"
  | "failed"
  | "loaded";

export type OllamaOptionValue = string | number | boolean | null;

export interface ModelSettings {
  defaultModel: string;
  style: TranslationStyle;
  honorificPolicy: HonorificPolicy;
  think: string | boolean;
  temperature: number;
  topP: number;
  contextWindow: number;
  maxTokens: number;
}

export interface TranslationOptions {
  model: string;
  promptVersion?: string;
  style: TranslationStyle;
  honorificPolicy: HonorificPolicy;
  think: string | boolean;
  useGlossary: boolean;
  useCache: boolean;
  preserveNames: boolean;
  temperature: number;
  topP: number;
  contextWindow: number;
  maxTokens: number;
}

export interface TranslationRequest {
  client_options?: TranslationOptions;
  text: string;
  model_name: string;
  prompt_version?: string;
  source_lang: SourceLanguage;
  target_lang: TargetLanguage;
  translate_scope: TranslateScope;
  page_index: number;
  style: TranslationStyle;
  honorific_policy: HonorificPolicy;
  preserve_names: boolean;
  use_glossary: boolean;
  use_cache: boolean;
  stream: false;
  think: string | boolean;
  options: {
    temperature: number;
    top_p: number;
    num_ctx: number;
    num_predict: number;
  };
}

export interface UrlTranslationRequest extends Omit<TranslationRequest, "text"> {
  url: string;
  view_mode: ViewerMode;
}

export interface PageTranslateRequest extends Omit<TranslationRequest, "text" | "translate_scope" | "page_index"> {
  force: boolean;
}

export interface TranslationChunkResponse {
  index: number;
  source_lang: string;
  target_lang: string;
  status: string;
}

export interface TranslationResponse {
  job_id: number;
  source_type: string;
  source_lang: string;
  target_lang: string;
  current_page_index: number;
  total_pages: number;
  has_next_page: boolean;
  translated_text: string;
  model: string;
  prompt_version: string;
  style: string;
  elapsed_ms: number;
  cache_hit: boolean;
  chunks: TranslationChunkResponse[];
}

export interface TranslationPage {
  index: number;
  id?: number;
  title?: string | null;
  sourceText: string;
  translatedText: string;
  status: PageStatus;
  totalChunks?: number;
  completedChunks?: number;
  failedChunks?: number;
  errorMessage?: string | null;
}

export interface TranslationChunk {
  id: number;
  pageId: number;
  pageIndex: number | null;
  index: number;
  sourceText: string;
  translatedText: string;
  status: string;
  retryCount: number;
  errorMessage: string | null;
}

export interface TranslationJob {
  id: string;
  jobId: number;
  title: string;
  inputMode: InputMode;
  sourceType: "pasted_text" | "url";
  sourceUrl?: string;
  sourceSite?: string;
  sourceAuthor?: string | null;
  sourceWorkId?: string | null;
  sourceFetchedAt?: string | null;
  sourceLang: SourceLanguage;
  targetLang: TargetLanguage;
  model: string;
  promptVersion: string;
  options: TranslationOptions;
  pages: TranslationPage[];
  chunks: TranslationChunk[];
  currentPageIndex: number;
  status: TranslationStatus;
  sourcePreview?: string;
  translatedPreview?: string;
  errorMessage?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface TranslationApiResult {
  response: TranslationResponse;
  job: TranslationJob;
}

export const DEFAULT_MODEL_SETTINGS: ModelSettings = {
  defaultModel: "gemma4:26b-a4b-it-q4_K_M",
  style: "webnovel",
  honorificPolicy: "preserve",
  think: false,
  temperature: 0.3,
  topP: 0.9,
  contextWindow: 8192,
  maxTokens: 4096,
};
